Python 3.12: tokenize.TokenError: ('unterminated string literal

Due to a change in Python 3.12, the template engine appears no longer able to properly parse templates.

For example, on Open Library, trying to render _any_ template appears to result in:
```
(...snipped for brevity -- full context further down the issue)
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 318, in get_tokens
web-1           |     for t in tokenize.generate_tokens(readline):
web-1           |   File "/usr/local/lib/python3.12/tokenize.py", line 543, in _generate_tokens_from_c_tokenizer
web-1           |     raise TokenError(msg, (e.lineno, e.offset)) from None
web-1           | tokenize.TokenError: ('unterminated string literal (detected at line 1)', (1, 4))
```

The difference can bee seen as follows:
*Python 3.11.2*
```python
(venv311) 9c6e9bd7c219% python
Python 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import web
>>> parser = web.template.Parser()
>>> parser.read_suite('<p><a href="$back">&larr; Back to Index</a></p>')
[<line: [t'<p><a href="', $back, t'">&larr; Back to Index</a></p>']>]
```

*Python 3.12.1*
```python
9c6e9bd7c219% python
Python 3.12.1 (main, Dec 19 2023, 03:09:45) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import web
>>> parser = web.template.Parser()
>>> parser.read_expr('<p><a href="$back">&larr; Back to Index</a></p>')
tokenize.TokenError: ('unterminated string literal (detected at line 1)', (1, 5))
```

Drilling down a bit one of the tokens, notably with an unmatched `"`, that is passed to `tokenize.generate_tokens()`:
*Python 3.11.2*
```python
In [11]: file = io.StringIO('$back">&larr; Back to Index</a></p>\n')

In [12]: for token in tokenize.generate_tokens(file.readline):
    ...:     print(token)
    ...: 
TokenInfo(type=60 (ERRORTOKEN), string='$', start=(1, 0), end=(1, 1), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=1 (NAME), string='back', start=(1, 1), end=(1, 5), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=60 (ERRORTOKEN), string='"', start=(1, 5), end=(1, 6), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=54 (OP), string='>', start=(1, 6), end=(1, 7), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=54 (OP), string='&', start=(1, 7), end=(1, 8), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=1 (NAME), string='larr', start=(1, 8), end=(1, 12), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=54 (OP), string=';', start=(1, 12), end=(1, 13), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=1 (NAME), string='Back', start=(1, 14), end=(1, 18), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=1 (NAME), string='to', start=(1, 19), end=(1, 21), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=1 (NAME), string='Index', start=(1, 22), end=(1, 27), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=54 (OP), string='<', start=(1, 27), end=(1, 28), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=54 (OP), string='/', start=(1, 28), end=(1, 29), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=1 (NAME), string='a', start=(1, 29), end=(1, 30), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=54 (OP), string='>', start=(1, 30), end=(1, 31), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=54 (OP), string='<', start=(1, 31), end=(1, 32), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=54 (OP), string='/', start=(1, 32), end=(1, 33), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=1 (NAME), string='p', start=(1, 33), end=(1, 34), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=54 (OP), string='>', start=(1, 34), end=(1, 35), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=4 (NEWLINE), string='\n', start=(1, 35), end=(1, 36), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=0 (ENDMARKER), string='', start=(2, 0), end=(2, 0), line='')
```

*Python 3.12.1*
```python
In [265]: file = io.StringIO('$back">&larr; Back to Index</a></p>\n')

In [266]: for token in tokenize.generate_tokens(file.readline):
     ...:     print(token)
     ...: 
TokenInfo(type=55 (OP), string='$', start=(1, 0), end=(1, 1), line='$back">&larr; Back to Index</a></p>\n')
TokenInfo(type=1 (NAME), string='back', start=(1, 1), end=(1, 5), line='$back">&larr; Back to Index</a></p>\n')
---------------------------------------------------------------------------
TokenError                                Traceback (most recent call last)
Cell In[266], line 1
----> 1 for token in tokenize.generate_tokens(file.readline):
      2     print(token)

File ~/.pyenv/versions/3.12.1/lib/python3.12/tokenize.py:543, in _generate_tokens_from_c_tokenizer(source, encoding, extra_tokens)
    541     raise e from None
    542 msg = _transform_msg(e.msg)
--> 543 raise TokenError(msg, (e.lineno, e.offset)) from None

TokenError: ('unterminated string literal (detected at line 1)', (1, 6))
```

It may not be significant, but whereas a lone `"` was formerly an `ERRORTOKEN`, now it's an `OP`. Either way, having an unmatched `"` seems to be a problem with `tokenize.generate_tokens()` now.

To the extent it is relevant, here are changes made to `tokenize` in Python 3.12: https://github.com/python/cpython/issues/111224#issuecomment-1776037598.

See also: internetarchive/openlibrary#8694


As promised, here's more context for the error:
```
web-1           | [2024-01-06 05:40:12 +0000] [10] [ERROR] Error handling request /
web-1           | Traceback (most recent call last):
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/gunicorn/workers/sync.py", line 136, in handle
web-1           |     self.handle_request(listener, req, client, addr)
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/gunicorn/workers/sync.py", line 179, in handle_request
web-1           |     respiter = self.wsgi(environ, resp.start_response)
web-1           |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/httpserver.py", line 258, in __call__
web-1           |     return self.app(environ, start_response)
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/openlibrary/openlibrary/core/middleware.py", line 46, in __call__
web-1           |     data = self.app(environ, new_start_response)
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/openlibrary/scripts/openlibrary-server", line 129, in wrapper
web-1           |     return app(environ, start_response)
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/application.py", line 315, in wsgi
web-1           |     result = self.handle_with_processors()
web-1           |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/application.py", line 286, in handle_with_processors
web-1           |     return process(self.processors)
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/application.py", line 283, in process
web-1           |     raise self.internalerror()
web-1           |           ^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/openlibrary/openlibrary/plugins/openlibrary/code.py", line 942, in internalerror
web-1           |     name = save_error()
web-1           |            ^^^^^^^^^^^^
web-1           |   File "/openlibrary/openlibrary/plugins/openlibrary/code.py", line 931, in save_error
web-1           |     error = web.safestr(web.djangoerror())
web-1           |                         ^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/debugerror.py", line 307, in djangoerror
web-1           |     djangoerror_r = Template(djangoerror_t, filename=__file__, filter=websafe)
web-1           |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 932, in __init__
web-1           |     code = self.compile_template(text, filename)
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 1003, in compile_template
web-1           |     code = Template.generate_code(
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 988, in generate_code
web-1           |     rootnode = parser.parse(text, filename)
web-1           |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 87, in parse
web-1           |     suite = self.read_suite(text)
web-1           |             ^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 175, in read_suite
web-1           |     section, text = self.read_section(text)
web-1           |                     ^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 119, in read_section
web-1           |     return self.read_block_section(text2, begin_indent)
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 477, in read_block_section
web-1           |     return self.create_block_node(keyword, stmt, block, begin_indent), text
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 481, in create_block_node
web-1           |     return self.statement_nodes[keyword](stmt, block, begin_indent)
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 692, in __init__
web-1           |     BlockNode.__init__(self, *a, **kw)
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 626, in __init__
web-1           |     self.suite = Parser().read_suite(block)
web-1           |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 175, in read_suite
web-1           |     section, text = self.read_section(text)
web-1           |                     ^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 119, in read_section
web-1           |     return self.read_block_section(text2, begin_indent)
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 477, in read_block_section
web-1           |     return self.create_block_node(keyword, stmt, block, begin_indent), text
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 481, in create_block_node
web-1           |     return self.statement_nodes[keyword](stmt, block, begin_indent)
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 626, in __init__
web-1           |     self.suite = Parser().read_suite(block)
web-1           |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 175, in read_suite
web-1           |     section, text = self.read_section(text)
web-1           |                     ^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 126, in read_section
web-1           |     return self.readline(text)
web-1           |            ^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 198, in readline
web-1           |     node, line = self.read_node(line)
web-1           |                  ^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 224, in read_node
web-1           |     return self.read_expr(text, escape=escape)
web-1           |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 362, in read_expr
web-1           |     simple_expr()
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 271, in simple_expr
web-1           |     extended_expr()
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 277, in extended_expr
web-1           |     lookahead = tokens.lookahead()
web-1           |                 ^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 338, in lookahead
web-1           |     self.items.append(self._next())
web-1           |                       ^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 343, in _next
web-1           |     return next(self.iteritems)
web-1           |            ^^^^^^^^^^^^^^^^^^^^
web-1           |   File "/home/openlibrary/.local/lib/python3.12/site-packages/web/template.py", line 318, in get_tokens
web-1           |     for t in tokenize.generate_tokens(readline):
web-1           |   File "/usr/local/lib/python3.12/tokenize.py", line 543, in _generate_tokens_from_c_tokenizer
web-1           |     raise TokenError(msg, (e.lineno, e.offset)) from None
web-1           | tokenize.TokenError: ('unterminated string literal (detected at line 1)', (1, 4))
```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -83,14 +83,13 @@ show-source = true
 target-version = "py38"
 
 [tool.ruff.mccabe]
-max-complexity = 26
+max-complexity = 20
 
 [tool.ruff.pylint]
 allow-magic-value-types = ["int", "str"]
 max-args = 9  # default is 5
 max-branches = 17  # default is 12
 max-returns = 8  # default is 6
-max-statements = 51  # default is 50
 
 [tool.codespell]
 ignore-words-list = "asend,gae"

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,2 +1,3 @@
 cheroot>=6.0.0
+more_itertools>=2.6
 multipart>=0.2.4

--- a/tests/test_template.py
+++ b/tests/test_template.py
@@ -4,6 +4,13 @@ import web
 from web.template import ExpressionNode, Parser, SecurityError, Template
 
 
+class TestItem:
+    __test__ = False  # silence collection warning from test framework
+
+    def __init__(self):
+        self.id = 12345
+
+
 class _TestResult:
     def __init__(self, t):
         self.t = t
@@ -47,6 +54,16 @@ class TemplateTest(unittest.TestCase):
         template = 'a="$foo" <p>'
         f = t(template, globals={"foo": "bar"})
         assert repr(f()) == "'a=\"bar\" <p>\\n'"
+
+    def test_accessor(self):
+        template = 'a="$foo.id"<p>'
+        f = t(template, globals={"foo": TestItem()})
+        assert repr(f()) == "'a=\"12345\"<p>\\n'"
+
+    def test_href(self):
+        template = '<a href="/del/$item.id">Delete</a>'
+        f = t(template, globals={"item": TestItem()})
+        assert repr(f()) == "'<a href=\"/del/12345\">Delete</a>\\n'"
 
 
 class TestParser(unittest.TestCase):

--- a/tests/test_template.py
+++ b/tests/test_template.py
@@ -1,7 +1,7 @@
 import unittest
 
 import web
-from web.template import SecurityError, Template
+from web.template import ExpressionNode, Parser, SecurityError, Template
 
 
 class _TestResult:
@@ -42,6 +42,51 @@ class TemplateTest(unittest.TestCase):
         tpl = "$print('blah')"
         f = t(tpl, globals={"print": lambda x: x})
         assert repr(f()) == "'blah\\n'"
+
+    def test_quotes(self):
+        template = 'a="$foo" <p>'
+        f = t(template, globals={"foo": "bar"})
+        assert repr(f()) == "'a=\"bar\" <p>\\n'"
+
+
+class TestParser(unittest.TestCase):
+    """
+    Test the Parser.
+
+    Tests functions from the Parser class as if the following template were loaded:
+
+    test_template = '''$def with (back, docs)
+    $var title: Index
+    <p><a href="$back">&larr; Back to Index</a></p>
+    <ul>
+    $for path, title in docs:
+        <li><a href="$path">$title</a></li>
+    </ul>
+    '''
+    """
+
+    def test_read_expr(self) -> None:
+        """
+        Test Parser.read_expr() with the `text` values it would get from
+        `Parser.read_node(), if processing `test_template`.
+        """
+        got = Parser().read_expr('back">&larr; Back to Index</a></p>\n')
+        expression_node = got[0]
+        assert isinstance(expression_node, ExpressionNode)
+        assert repr(expression_node) == "$back"
+        assert got[1] == '">&larr; Back to Index</a></p>\n'
+
+        got = Parser().read_expr('path">$title</a></li>\n')
+        expression_node = got[0]
+        assert isinstance(expression_node, ExpressionNode)
+        assert repr(expression_node) == "$path"
+        assert got[1] == '">$title</a></li>\n'
+
+        got = Parser().read_expr("title</a></li>\n")
+        expression_node = got[0]
+        assert isinstance(expression_node, ExpressionNode)
+        assert repr(expression_node) == "$title"
+        assert got[1] == "</a></li>\n"
 
 
 class TestRender:

--- a/web/browser.py
+++ b/web/browser.py
@@ -1,6 +1,7 @@
 """Browser to test web applications.
 (from web.py)
 """
+
 import os
 import webbrowser
 from http.cookiejar import CookieJar
@@ -46,7 +47,7 @@ class Browser:
 
     def build_opener(self):
         """Builds the opener using (urllib2/urllib.request).build_opener.
-        Subclasses can override this function to prodive custom openers.
+        Subclasses can override this function to provide custom openers.
         """
         return urllib_build_opener()
 

--- a/web/contrib/template.py
+++ b/web/contrib/template.py
@@ -1,6 +1,7 @@
 """
 Interface to various templating engines.
 """
+
 import os.path
 
 __all__ = ["render_cheetah", "render_genshi", "render_mako", "cache"]

--- a/web/db.py
+++ b/web/db.py
@@ -2,6 +2,7 @@
 Database API
 (part of web.py)
 """
+
 import ast
 import datetime
 import os

--- a/web/net.py
+++ b/web/net.py
@@ -3,7 +3,6 @@ Network Utilities
 (from web.py)
 """
 
-
 import datetime
 import re
 import socket

--- a/web/template.py
+++ b/web/template.py
@@ -31,8 +31,10 @@ Grammar:
 import ast
 import builtins
 import glob
+import itertools
 import os
 import sys
+import token
 import tokenize
 from functools import partial
 
@@ -331,13 +333,22 @@ class Parser:
             try:
                 yield from tokenize_text(text)
             except tokenize.TokenError as e:
-                # Python 3.12+ compatibility: add a `"` suffix to terminate the string literal.
-                # See https://github.com/webpy/webpy/issues/784.
                 if "unterminated string literal" in str(e):
-                    fixed_text = text + '"'
-                    yield from tokenize_text(fixed_text)
+                    # Things like unterminated string literals or EOF in multi-line literals will raise exceptions.
+                    # Tokenize the error free portion, then return an error token with the rest of the text.
+                    error_pos = e.args[1][1] - 1
+                    fixed_text = text[0:error_pos]
+                    yield from itertools.chain(
+                        tokenize_text(fixed_text),
+                        error_token_generator(text, error_pos, len(text)),
+                    )
                 else:
-                    raise  # Re-raise the exception if it's not due to an unterminated string literal
+                    raise
+
+        def error_token_generator(text, start, end):
+            yield storage(
+                type=token.ERRORTOKEN, value=text[start:], begin=start, end=end
+            )
 
         class BetterIter:
             """Iterator like object with 2 support for 2 look aheads."""

--- a/web/template.py
+++ b/web/template.py
@@ -38,6 +38,8 @@ import token
 import tokenize
 from functools import partial
 
+from more_itertools import peekable
+
 from .net import websafe
 from .utils import re_compile, safestr, safeunicode, storage
 from .webapi import config
@@ -273,10 +275,10 @@ class Parser:
             extended_expr()
 
         def identifier():
-            next(tokens)
+            return next(tokens)
 
         def extended_expr():
-            lookahead = tokens.lookahead()
+            lookahead = tokens.peek()
             if lookahead is None:
                 return
             elif lookahead.value == ".":
@@ -290,7 +292,7 @@ class Parser:
         def attr_access():
             from token import NAME  # python token constants
 
-            if tokens.lookahead2().type == NAME:
+            if tokens[1].type == NAME:
                 next(tokens)  # consume dot
                 identifier()
                 extended_expr()
@@ -299,7 +301,7 @@ class Parser:
             begin = next(tokens).value
             end = parens[begin]
             while True:
-                if tokens.lookahead().value in parens:
+                if tokens.peek().value in parens:
                     paren_expr()
                 else:
                     t = next(tokens)
@@ -350,39 +352,22 @@ class Parser:
                 type=token.ERRORTOKEN, value=text[start:], begin=start, end=end
             )
 
-        class BetterIter:
-            """Iterator like object with 2 support for 2 look aheads."""
+        class peekable2(peekable):
+            """
+            A peekable class which caches the last item returned by next()
+            """
 
-            def __init__(self, items):
-                self.iteritems = iter(items)
-                self.items = []
-                self.position = 0
+            def __init__(self, iterable):
+                super().__init__(iterable)
                 self.current_item = None
 
-            def lookahead(self):
-                if len(self.items) <= self.position:
-                    self.items.append(self._next())
-                return self.items[self.position]
-
-            def _next(self):
-                try:
-                    return next(self.iteritems)
-                except StopIteration:
-                    return None
-
-            def lookahead2(self):
-                if len(self.items) <= self.position + 1:
-                    self.items.append(self._next())
-                return self.items[self.position + 1]
-
             def __next__(self):
-                self.current_item = self.lookahead()
-                self.position += 1
+                self.current_item = super().__next__()
                 return self.current_item
 
-        tokens = BetterIter(get_tokens(text))
+        tokens = peekable2(get_tokens(text))
 
-        if tokens.lookahead().value in parens:
+        if tokens.peek().value in parens:
             paren_expr()
         else:
             simple_expr()

--- a/web/template.py
+++ b/web/template.py
@@ -335,17 +335,14 @@ class Parser:
             try:
                 yield from tokenize_text(text)
             except tokenize.TokenError as e:
-                if "unterminated string literal" in str(e):
-                    # Things like unterminated string literals or EOF in multi-line literals will raise exceptions.
-                    # Tokenize the error free portion, then return an error token with the rest of the text.
-                    error_pos = e.args[1][1] - 1
-                    fixed_text = text[0:error_pos]
-                    yield from itertools.chain(
-                        tokenize_text(fixed_text),
-                        error_token_generator(text, error_pos, len(text)),
-                    )
-                else:
-                    raise
+                # Things like unterminated string literals or EOF in multi-line literals will raise exceptions
+                # tokenize the error free portion, then return an error token with the rest of the text
+                error_pos = e.args[1][1] - 1
+                fixed_text = text[0:error_pos]
+                yield from itertools.chain(
+                    tokenize_text(fixed_text),
+                    error_token_generator(text, error_pos + 1, len(text)),
+                )
 
         def error_token_generator(text, start, end):
             yield storage(

--- a/web/template.py
+++ b/web/template.py
@@ -242,7 +242,7 @@ class Parser:
         line, text = splitline(text)
         return StatementNode(line.strip() + "\n"), text
 
-    def read_expr(self, text, escape=True):
+    def read_expr(self, text, escape=True):  # noqa: C901, PLR0915
         """Reads a python expression from the text and returns the expression and remaining text.
 
         expr -> simple_expr | paren_expr

--- a/web/template.py
+++ b/web/template.py
@@ -306,23 +306,38 @@ class Parser:
 
         parens = {"(": ")", "[": "]", "{": "}"}
 
-        def get_tokens(text):
+        def get_tokens(text: str):
             """tokenize text using python tokenizer.
             Python tokenizer ignores spaces, but they might be important in some cases.
             This function introduces dummy space tokens when it identifies any ignored space.
             Each token is a storage object containing type, value, begin and end.
             """
-            i = iter([text])
-            readline = lambda: next(i)
-            end = None
-            for t in tokenize.generate_tokens(readline):
-                t = storage(type=t[0], value=t[1], begin=t[2], end=t[3])
-                if end is not None and end != t.begin:
-                    _, x1 = end
-                    _, x2 = t.begin
-                    yield storage(type=-1, value=text[x1:x2], begin=end, end=t.begin)
-                end = t.end
-                yield t
+
+            def tokenize_text(input_text):
+                i = iter([input_text])
+                readline = lambda: next(i)
+                end = None
+                for t in tokenize.generate_tokens(readline):
+                    t = storage(type=t[0], value=t[1], begin=t[2], end=t[3])
+                    if end is not None and end != t.begin:
+                        _, x1 = end
+                        _, x2 = t.begin
+                        yield storage(
+                            type=-1, value=input_text[x1:x2], begin=end, end=t.begin
+                        )
+                    end = t.end
+                    yield t
+
+            try:
+                yield from tokenize_text(text)
+            except tokenize.TokenError as e:
+                # Python 3.12+ compatibility: add a `"` suffix to terminate the string literal.
+                # See https://github.com/webpy/webpy/issues/784.
+                if "unterminated string literal" in str(e):
+                    fixed_text = text + '"'
+                    yield from tokenize_text(fixed_text)
+                else:
+                    raise  # Re-raise the exception if it's not due to an unterminated string literal
 
         class BetterIter:
             """Iterator like object with 2 support for 2 look aheads."""

--- a/web/test.py
+++ b/web/test.py
@@ -1,6 +1,7 @@
 """test utilities
 (part of web.py)
 """
+
 import doctest
 import sys
 import unittest

